### DISCUSSION TITLE
Best approach for testing bot locally while not affecting production?

### DISCUSSION THREAD

**Discussion by jsjoeio** [telegraf/telegraf#1482] [Q&A] [answered]
What's the recommended approach for testing my bot locally while keeping my bot working in production?

I know I can run it locally, expose with `ngrok` and set the webhook to the ngrok url but worried about the down time. Only other thing I can think of is creating a separate bot and changing out the token/webhook locally to test things. Is there a better way?

_cross-posted in the Telegram group_

**Comment by jsjoeio** [comment, marked as answer]
> You could still test your functions with unit and integration tests without starting the bot. Alternatively, if by "test", you meant you wanted to try it out manually from your telegram client, then the only way is to use a different token and webhook url.
— @KnightNiwrem 


Seems like the suggested approach is:
- write unit tests for functions
- write integration tests to test one level higher
- for e2e tests (testing with the client), use a separate token and webhook url

**Reply by jsjoeio** [reply]
Looks like there is a library which could help with testing as well: [`telegram-test-api`](https://github.com/jehy/telegram-test-api)

**Comment by jsstrn** [comment]
My approach is to have two separate bots. One for local and one for production. 

When I run my server locally, I use the bot token that's meant for local purposes only. When I deploy my code to production, it uses a different bot token (which is a different bot) so I never affect anything in production. 

Essentially, you'd want a separate bot for each environment (e.g. dev, testing, staging, prod).

**Reply by jsjoeio** [reply]
Awesome! Thanks for sharing that @jsstrn. That seems like a great approach and pretty easy to set up.
